src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.MoveFileCodeAction.cs

```diff
@@ -0,0 +1,189 @@
+                            : ImmutableArray.Create(parts, index, parts.Length - index);
+                        var newCurrentFolder = currentFolder.Add(matchingFolderInfo.Name);
+                        builder.AddRange(FindCandidateFolders(matchingFolderInfo, newParts, newCurrentFolder));
+                    }
+                }
+
+                // Make sure we always have the default path as an available option to the user 
+                // (which might have been found by the search above, therefore the check here)
+                // For example, if the target namespace is "A.B.C.D", and there's folder <ROOT>\A.B\,
+                // the search above would only return "<ROOT>\A.B\C\D". We'd want to provide 
+                // "<ROOT>\A\B\C\D" as the default path.
+                var defaultPathBasedOnCurrentFolder = currentFolder.AddRange(parts);
+                if (builder.All(folders => !folders.SequenceEqual(defaultPathBasedOnCurrentFolder, PathUtilities.Comparer)))
+                {
+                    builder.Add(defaultPathBasedOnCurrentFolder);
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            private class FolderInfo
+            {
+                private Dictionary<string, FolderInfo> _childFolders;
+
+                public string Name { get; }
+
+                public IReadOnlyDictionary<string, FolderInfo> ChildFolders => _childFolders;
+
+                private FolderInfo(string name)
+                {
+                    Name = name;
+                    _childFolders = new Dictionary<string, FolderInfo>(StringComparer.Ordinal);
+                }
+
+                private void AddFolder(IEnumerable<string> folder)
+                {
+                    if (!folder.Any())
+                    {
+                        return;
+                    }
+
+                    var firstFolder = folder.First();
+                    if (!_childFolders.TryGetValue(firstFolder, out var firstFolderInfo))
+                    {
+                        firstFolderInfo = new FolderInfo(firstFolder);
+                        _childFolders[firstFolder] = firstFolderInfo;
+                    }
+
+                    firstFolderInfo.AddFolder(folder.Skip(1));
+                }
+
+                // TODO: 
+                // Since we are getting folder data from documents, only non-empty folders 
+                // in the project are discovered. It's possible to get complete folder structure
+                // from VS but it requires UI thread to do so. We might want to revisit this later.
+                public static FolderInfo CreateFolderHierarchyForProject(Project project)
+                {
+                    var handledFolders = new HashSet<string>(StringComparer.Ordinal);
+
+                    var rootFolderInfo = new FolderInfo("<ROOT>");
+                    foreach (var document in project.Documents)
```

Comment:
so, I guess we are okay to have race here? since project is a snapshot and it might not actually a reality.

Reply:
I believe so.
1. If user doesn't see what they need from the list of actions because of race, they can simplt trigger the refactoring again.
2. If user wants to choose a folder which is no longer there, they can do so without any problem